nexus/db-queries/src/transaction_retry.rs

```diff
@@ -0,0 +1,341 @@
+        Func: Fn(async_bb8_diesel::Connection<crate::db::DbConnection>) -> Fut
+            + Send
+            + Sync,
+    {
+        conn.transaction_async_with_retry(f, self.as_callback()).await
+    }
+
+    // Called upon retryable transaction failure.
+    //
+    // This function:
+    // - Appends a metric identifying the duration of the transaction operation
+    // - Performs a random (uniform) backoff (limited to less than 50 ms)
+    // - Returns "true" if the transaction should be restarted
+    async fn retry_callback(&self) -> bool {
+        // Look at the current attempt and start time so we can log this
+        // information before we start sleeping.
+        let (start, attempt) = {
+            let inner = self.inner.lock().unwrap();
+            (inner.start, inner.attempts)
+        };
+
+        let latency = (Utc::now() - start)
+            .to_std()
+            .unwrap_or(Duration::ZERO)
+            .as_secs_f64();
+
+        let _ = self.producer.append(
+            &DatabaseTransaction { name: self.name.into() },
+            &RetryData { latency, attempt },
+        );
+
+        // This backoff is not exponential, but I'm not sure we actually want
+        // that much backoff here. If we're repeatedly failing, it would
+        // probably be better to fail the operation, at which point Oximeter
+        // will keep track of the failing transaction and identify that it's a
+        // high-priority target for CTE conversion.
+        let duration = {
+            let mut rng = thread_rng();
+            rng.gen_range(MIN_RETRY_BACKOFF..MAX_RETRY_BACKOFF)
+        };
+        tokio::time::sleep(duration).await;
+
+        // Now that we've finished sleeping, reset the timer and bump the number
+        // of attempts we've tried.
+        let inner = self.inner.lock().unwrap().tick();
+        return inner.attempts < MAX_RETRY_ATTEMPTS;
+    }
+
+    /// Converts this function to a retryable callback that can be used from
+    /// "transaction_async_with_retry".
+    pub(crate) fn as_callback(
+        self,
+    ) -> impl Fn() -> futures::future::BoxFuture<'static, bool> {
+        let r = Arc::new(self);
+        move || {
+            let r = r.clone();
+            Box::pin(async move { r.retry_callback().await })
+        }
+    }
+}
+
+impl oximeter::Producer for Producer {
+    fn produce(
+        &mut self,
+    ) -> Result<Box<dyn Iterator<Item = Sample> + 'static>, MetricsError> {
+        let samples = std::mem::take(&mut *self.samples.lock().unwrap());
+        Ok(Box::new(samples.into_iter()))
+    }
+}
+
+/// Helper utility for passing non-retryable errors out-of-band from
+/// transactions.
+///
+/// Transactions prefer to act on the `diesel::result::Error` type,
+/// but transaction users may want more meaningful error types.
+/// This utility helps callers safely propagate back Diesel errors while
+/// retaining auxiliary error info.
+pub struct OptionalError<E>(Arc<Mutex<Option<E>>>);
```

Comment:
I added this as a helper to simplify a bunch of the spots where we observe a diesel error, consider bailing immediately if it's retryable, and otherwise set an out-of-band error while returning `RollbackTransaction`  from the Diesel closure.